### ISSUE TITLE
Type middy.MiddlewareFn doesn't exist

### ISSUE THREAD

**Issue by jamiechong** [middyjs/middy#1147]
I'm using v5 of middy and reading these docs on custom middleware. https://middy.js.org/docs/writing-middlewares/with-typescript

But middy.MiddlewareFn isn't available.

**Comment by willfarrell** [comment]
I see it in @middy/core/index.d.ts
```
declare namespace middy {
  export {
    Request,
    PluginHook,
    PluginHookWithMiddlewareName,
    PluginObject,
    MiddlewareFn,
    MiddlewareObj,
    MiddyfiedHandler
  }
}

export default middy
```

I'm not sure why it wouldn't be showing up for you.

**Comment by jamiechong** [comment]
@willfarrell  - VSCode doesn't seem to think so. 

<img width="924" alt="Screenshot 2023-12-03 at 10 26 22 AM" src="https://github.com/middyjs/middy/assets/981664/30fe1ea4-603c-4916-ac51-9fbe52b34353">
<img width="867" alt="Screenshot 2023-12-03 at 10 26 38 AM" src="https://github.com/middyjs/middy/assets/981664/fc767661-7164-4337-a946-2722af96e49c">

**Comment by meganmajewski** [comment]
I'm having this same issue with /@middy/http-json-body-parser/package.json

> When importing and using @middy/core in my Node.js AWS Lambda application, I am running into the following error:
> 
> `Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in "<path>/node_modules/@middy/core/package.json"` This prevents being able to import and initialize middy properly.
> 
> Steps to Reproduce
> 
>     * Create Node.js handler that imports @middy/core
> 
>     * Initialize middy middleware with middy()
> 
>     * Build and run application
> 
>     * Notice error when middy is imported
> 
> 
> Cause Based on the error, it seems there needs to be an "exports" field defined in @middy/core's package.json to point to importable paths within the module.
> 
> When this "exports" mapping is not defined, Node is unable to resolve the import request properly at runtime.

**Comment by Valvide** [comment]
@MeganMajewski, I resolved the issue by updating the CommonJS usage as mentioned in the updated documentation for Middy. They've made changes from the previous usage, so review the migration guide from v4 to v5 for more details.

**Comment by jobuit** [comment]
Same error Here! @Valvide how do you resolve the issue ? 
I was looking in the upgrade information but I didn't find anything that help me.
https://middy.js.org/docs/upgrade/4-5
<img width="687" alt="image" src="https://github.com/middyjs/middy/assets/28201079/b4a522fd-cafa-4dd8-9ae6-1695203163d8">

**Comment by david-eolas** [comment]
Having this same issue with @middy/http-json-body-parser
The project is set with Node 20, Yarn 4, serverless 3.38 and middy 5.2.3

Error:
```
Cannot load "serverless.ts": Initialization error: Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in /Users/espi/workspace/backend/my-files-service/node_modules/@middy/http-json-body-parser/package.json
    at new NodeError (node:internal/errors:405:5)
    at exportsNotFound (node:internal/modules/esm/resolve:359:10)
    at packageExportsResolve (node:internal/modules/esm/resolve:639:13)
    at resolveExports (node:internal/modules/cjs/loader:567:36)
    at Function.Module._findPath (node:internal/modules/cjs/loader:636:31)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1063:27)
    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue (/Users/espi/workspace/backend/my-files-service/node_modules/@cspotcode/source-map-support/source-map-support.js:811:30)
    at Function.Module._resolveFilename (/Users/espi/workspace/backend/my-files-service/node_modules/tsconfig-paths/src/register.ts:115:36)
   .........
```

**Comment by lmammino** [comment]
Hello all, sorry for the delay in addressing this. 

I have tried to replicate the issue and I might have a possible solution in a [branch](https://github.com/middyjs/middy/tree/fix-1147).

Could you try to install `@middy/core` from this branch and see if that solves your issues? You should be able to do that with:

```bash
npm install 'https://gitpkg.now.sh/middyjs/middy/packages/core?fix-1147'
```

or by setting the following in your `package.json` and then running `npm install`:

```json
{
  "dependencies": {
    "@middy/core": "https://gitpkg.now.sh/middyjs/middy/packages/core?fix-1147"
  }
}
```

It might also be useful to see what your `tsconfig.json` looks like. Here's my current one for reference:

```json
{
    "compilerOptions": {
      "target": "es2020",
      "strict": true,
      "preserveConstEnums": true,
      "noEmit": true,
      "sourceMap": false,
      "module":"es2015",
      "moduleResolution":"node",
      "esModuleInterop": true, 
      "skipLibCheck": true,
      "forceConsistentCasingInFileNames": true,  
    },
    "exclude": ["node_modules", "**/*.test.ts"]
  }
```

If this helps fix your issue, we can create a PR, do a more formal review, and make sure the changes land in a new release.

Thanks


> [!NOTE]  
> Please note that we are not TypeScript experts and maintaining correct types and typescript configuration and docs is always something we struggle with. If someone wants to help to maintain that part more proactively we would love that.

**Comment by mkarsene** [comment]
@lmammino That didnt work for me...

**Comment by jobuit** [comment]
It looks that works! but I'm getting this typescript error now
<img width="572" alt="image" src="https://github.com/middyjs/middy/assets/28201079/1fbeaacb-d90b-4da9-82c5-6493717a4bac">
<img width="297" alt="image" src="https://github.com/middyjs/middy/assets/28201079/60de74c6-dabd-4aab-af9d-4f44bc8378d8">
<img width="690" alt="image" src="https://github.com/middyjs/middy/assets/28201079/7963ee96-ddec-4aa7-b1f4-2432313e6a35">

**Comment by david-eolas** [comment]
@lmammino Same error from my side:
`Cannot load "serverless.ts": Initialization error: Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in /Users/espi/workspace/backend/my-files-service/node_modules/@middy/http-json-body-parser/package.json`

**Comment by lmammino** [comment]
> No "exports" main defined in .../node_modules/@middy/http-json-body-parser/package.json

That's interesting because I haven't changed that package (I only touched `@middy/core`)... So I am quite puzzled by this! 🤔 

Either way... thanks to @mkarsene, @david-eolas, and @jobuit for testing this change. Could you be so kind to share your TSConfig or, even better, a minimal reproducible example that I can use as a starting point to try to resolve the issue. I am afraid that different versions of typescript and different TSConfig files might lead to very different results, so we would just keep guessing if we don't work on specific shared examples...

**Comment by jonanpogi** [comment]
> > No "exports" main defined in .../node_modules/@middy/http-json-body-parser/package.json
> 
> That's interesting because I haven't changed that package (I only touched `@middy/core`)... So I am quite puzzled by this! 🤔
> 
> Either way... thanks to @mkarsene, @david-eolas, and @jobuit for testing this change. Could you be so kind to share your TSConfig or, even better, a minimal reproducible example that I can use as a starting point to try to resolve the issue. I am afraid that different versions of typescript and different TSConfig files might lead to very different results, so we would just keep guessing if we don't work on specific shared examples...

Hi,

I know I'm a bit late, but I also experienced this issue right now on my aws lambda function written in typescript,

here's my tsconfig.json

{
  "compilerOptions": {
    "target": "ES2015",
    "module": "CommonJS",
    "lib": ["ES6"],
    "allowJs": true,
    "outDir": "build",
    "rootDir": "src",
    "strict": true,
    "noImplicitAny": true,
    "esModuleInterop": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "sourceMap": true,
    "forceConsistentCasingInFileNames": true,
    "skipLibCheck": true
  },
  "include": ["src/**/*.ts"],
  "exclude": ["node_modules"]
}

**Comment by willfarrell** [comment]
@RickDir Middy v5 doesn't support CommonJS, v4 does if you're unable to update to ESM yet.

**Comment by RickDir** [comment]
> @RickDir Middy v5 doesn't support CommonJS, v4 does if you're unable to update to ESM yet.

Thanks @willfarrell  Testing with ESM exposed the install instructions I got were lacking as well (throws: Cannot find package '@middy/util') I’ll have to dive deeper into the documentation to make sure I'm starting off correctly. Thanks again

**Comment by micheledel** [comment]
Hi all,

someone has resolved the issue?

I'm encountering it with middy/core 5.3.1

**Comment by refjosh** [comment]
> @RickDir Middy v5 doesn't support CommonJS, v4 does if you're unable to update to ESM yet.

Installing `v4.7.0` does resolve the issue on my end

**Comment by willfarrell** [comment]
We think this should be resolved in 5.3.3. Please reopen if now

**Comment by gbisaga** [comment]
It is not resolved in 5.3.3, at least for my situation. However, I've been able to get around it. I figure this might act as a clue to the ultimate fix. The key was that it works fine, except when I do something like this:

```
function prepare(fn) {
  return middy(fn).(add a bunch of middleware here);
}
const portalsGetList = prepare(fn);

console.log(portalsGetList.name);
```

When I get rid of the direct reference and `.name`, it builds. Otherwise I am getting:

```
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:995:15)
    at Function.Module._resolveFilename.sharedData.moduleResolveFilenameHook.installedValue [as _resolveFilename] (C:\src\vkd\vermontkidsdata-app\node_modules\@cspotcode\source-map-support\source-map-support.js:811:30)
    at Function.Module._load (node:internal/modules/cjs/loader:841:27)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (C:\src\vkd\vermontkidsdata-app\lib\portals-functions.ts:6:1)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module.m._compile (C:\src\vkd\vermontkidsdata-app\node_modules\ts-node\src\index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Object.require.extensions.<computed> [as .ts] (C:\src\vkd\vermontkidsdata-app\node_modules\ts-node\src\index.ts:1621:12) {
  code: 'MODULE_NOT_FOUND',
```

**Comment by roastdev** [comment]
I'm running into the same thing on the latest version of Middy

`No "exports" main defined in /var/task/node_modules/@middy/core/package.json.`

My Lambda function does a simple export like this and we're using ESM now

  `"@middy/core": "^5.3.5"`

**Comment by colmeye** [comment]
I am also encountering this issue. I've tried `5.3.2`, `5.3.3`, and `5.4.0` with no luck.
`Cannot load "serverless.ts": Initialization error: Error: No "exports" main defined in pathstuff/packages/shared-api/node_modules/@middy/core/package.json`

**Comment by brucePedroGomes** [comment]
I encountered the same issue in the latest version, and resolved it by reverting to '@middy/core': '^4.7.0'.

**Comment by Gribbs** [comment]
Same deal with  "@middy/core": "^5.5.0"
